File: gui/ticker_validator.py
"""
股票代码验证模块

提供股票代码格式验证和名称查询功能。
支持 A股、港股、美股。
"""

import re
from typing import Tuple


def validate_a_share_code(code: str) -> bool:
    """
    验证 A 股代码格式

    规则:
    - 6位数字
    - 上海: 600xxx, 601xxx, 603xxx, 605xxx, 688xxx (科创板)
    - 深圳: 000xxx, 001xxx, 002xxx (中小板), 300xxx (创业板)
    """
    code = code.strip()
    if not re.match(r"^\d{6}$", code):
        return False

    prefix = code[:3]
    valid_prefixes = ["600", "601", "603", "605", "688", "000", "001", "002", "300"]
    return prefix in valid_prefixes


def validate_us_stock_code(code: str) -> bool:
    """
    验证美股代码格式

    规则:
    - 1-5 个大写字母
    - 可选: 带 . 的类别标识 (如 BRK.B, BRK-B)
    """
    code = code.strip().upper()
    # 标准股票代码: 1-5 个字母
    if re.match(r"^[A-Z]{1,5}$", code):
        return True
    # 带类别的代码: 如 BRK.B, BRK-B
    if re.match(r"^[A-Z]{1,4}[.\-][A-Z]$", code):
        return True
    return False


def validate_hk_stock_code(code: str) -> bool:
    """
    验证港股代码格式

    规则:
    - 4-5 位数字 + .HK 后缀
    - 例如: 0700.HK, 9988.HK, 00005.HK
    """
    code = code.strip().upper()
    # 标准格式: 数字.HK
    if re.match(r"^\d{4,5}\.HK$", code):
        return True
    # 也接受纯数字 (自动补 .HK)
    if re.match(r"^\d{1,5}$", code):
        return True
    return False


def normalize_ticker(code: str, market: str) -> str:
    """
    标准化股票代码

    Args:
        code: 原始代码
        market: 市场类型

    Returns:
        标准化后的代码
    """
    code = code.strip()

    if market == "US":
        return code.upper()
    elif market == "HK":
        code = code.upper()
        # 去除 .HK 后缀以便处理
        clean_code = code.replace(".HK", "")
        # 补齐前导零 (如 700 -> 0700)
        if len(clean_code) < 4:
            clean_code = clean_code.zfill(4)
        return f"{clean_code}.HK"
    else:
        # A股保持原样
        return code


def get_stock_name_a_share(ticker: str) -> Tuple[bool, str]:
    """
    获取 A 股股票名称

    Args:
        ticker: A股代码 (6位数字)

    Returns:
        (success, name_or_error)
    """
    try:
        import akshare as ak
        info = ak.stock_individual_info_em(symbol=ticker)
        name = info[info['item'] == '股票简称']['value'].values[0]
        return True, name
    except ImportError:
        # akshare 未安装，返回通用成功
        return True, ""
    except Exception as e:
        return False, f"未找到该股票: {ticker}"


def get_stock_name_hk(ticker: str) -> Tuple[bool, str]:
    """
    获取港股股票名称

    Args:
        ticker: 港股代码 (如 0700.HK)

    Returns:
        (success, name_or_error)
    """
    try:
        import yfinance as yf
        stock = yf.Ticker(ticker)
        info = stock.info
        name = info.get("longName") or info.get("shortName") or ""
        if name:
            return True, name
        return True, ""  # 可能没有名称但代码有效
    except ImportError:
        return True, ""
    except Exception as e:
        return False, f"未找到该股票: {ticker}"


def get_stock_name_us(ticker: str) -> Tuple[bool, str]:
    """
    获取美股股票名称

    Args:
        ticker: 美股代码 (如 AAPL)

    Returns:
        (success, name_or_error)
    """
    try:
        import yfinance as yf
        stock = yf.Ticker(ticker)
        info = stock.info
        name = info.get("longName") or info.get("shortName") or ""
        if name:
            return True, name
        return True, ""  # 可能没有名称但代码有效
    except ImportError:
        return True, ""
    except Exception as e:
        return False, f"未找到该股票: {ticker}"


def validate_and_get_name(ticker: str, market: str) -> Tuple[bool, str, str]:
    """
    验证股票代码并获取名称

    Args:
        ticker: 股票代码
        market: 市场类型 (A-share, US, HK)

    Returns:
        (is_valid, normalized_ticker, stock_name_or_error)
    """
    ticker = ticker.strip()

    if not ticker:
        return False, "", "请输入股票代码"

    # 格式验证
    if market == "A-share":
        if not validate_a_share_code(ticker):
            return False, "", "A股代码应为6位数字（如 600519、000001）"
        normalized = normalize_ticker(ticker, market)
        success, name = get_stock_name_a_share(normalized)
        if success:
            return True, normalized, name
        return False, "", name

    elif market == "US":
        if not validate_us_stock_code(ticker):
            return False, "", "美股代码应为1-5位字母（如 AAPL、TSLA）"
        normalized = normalize_ticker(ticker, market)
        success, name = get_stock_name_us(normalized)
        if success:
            return True, normalized, name
        return False, "", name

    elif market == "HK":
        if not validate_hk_stock_code(ticker):
            return False, "", "港股代码应为4-5位数字（如 0700、9988）"
        normalized = normalize_ticker(ticker, market)
        success, name = get_stock_name_hk(normalized)
        if success:
            return True, normalized, name
        return False, "", name

    return False, "", f"不支持的市场类型: {market}"


def validate_ticker_format(ticker: str, market: str) -> Tuple[bool, str, str]:
    """
    仅验证股票代码格式（不查询名称，速度快）

    Args:
        ticker: 股票代码
        market: 市场类型 (A-share, US, HK)

    Returns:
        (is_valid, normalized_ticker, error_message)
    """
    ticker = ticker.strip()

    if not ticker:
        return False, "", "请输入股票代码"

    if market == "A-share":
        if not validate_a_share_code(ticker):
            return False, "", "A股代码应为6位数字（如 600519、000001）"
        return True, normalize_ticker(ticker, market), ""

    elif market == "US":
        if not validate_us_stock_code(ticker):
            return False, "", "美股代码应为1-5位字母（如 AAPL、TSLA）"
        return True, normalize_ticker(ticker, market), ""

    elif market == "HK":
        if not validate_hk_stock_code(ticker):
            return False, "", "港股代码应为4-5位数字（如 0700、9988）"
        return True, normalize_ticker(ticker, market), ""

    return False, "", f"不支持的市场类型: {market}"

File: web/src/services/ticker-validator.ts
/**
 * Ticker Validator - Stock Code Validation Service
 *
 * Validates stock codes and retrieves stock names via HTTP APIs.
 * - A-Share: Eastmoney HTTP API
 * - US/HK: Yahoo Finance API
 */

import type { Market } from '../types'

interface ValidationResult {
  valid: boolean
  normalized: string
  name: string
  error: string
}

// A-Share code format validation
function validateAShareFormat(code: string): boolean {
  code = code.trim()
  if (!/^\d{6}$/.test(code)) {
    return false
  }

  const prefix = code.substring(0, 3)
  const validPrefixes = ['600', '601', '603', '605', '688', '000', '001', '002', '300']
  return validPrefixes.includes(prefix)
}

// US stock code format validation
function validateUSFormat(code: string): boolean {
  code = code.trim().toUpperCase()
  // Standard ticker: 1-5 letters
  if (/^[A-Z]{1,5}$/.test(code)) {
    return true
  }
  // Class-based ticker: e.g. BRK.B, BRK-B
  if (/^[A-Z]{1,4}[.\-][A-Z]$/.test(code)) {
    return true
  }
  return false
}

// HK stock code format validation
function validateHKFormat(code: string): boolean {
  code = code.trim().toUpperCase()
  // Standard format: number.HK
  if (/^\d{4,5}\.HK$/.test(code)) {
    return true
  }
  // Also accept plain number (auto-append .HK)
  if (/^\d{1,5}$/.test(code)) {
    return true
  }
  return false
}

// Normalize ticker code
function normalizeTicker(code: string, market: Market): string {
  code = code.trim()

  if (market === 'US') {
    return code.toUpperCase()
  } else if (market === 'HK') {
    code = code.toUpperCase()
    const cleanCode = code.replace('.HK', '')
    const paddedCode = cleanCode.padStart(4, '0')
    return `${paddedCode}.HK`
  } else {
    // A-Share: keep as-is
    return code
  }
}

/**
 * Validate A-Share code and get name
 * Uses Eastmoney HTTP API
 */
async function validateAShare(ticker: string): Promise<ValidationResult> {
  const normalized = ticker

  // Try Shanghai (secid=1.xxx)
  try {
    const url = `https://push2.eastmoney.com/api/qt/stock/get?secid=1.${ticker}&fields=f58`
    const res = await fetch(url, { signal: AbortSignal.timeout(5000) })
    const data = await res.json() as any

    if (data.data?.f58) {
      return {
        valid: true,
        normalized,
        name: data.data.f58,
        error: '',
      }
    }
  } catch {
    // Continue trying Shenzhen
  }

  // Try Shenzhen (secid=0.xxx)
  try {
    const url = `https://push2.eastmoney.com/api/qt/stock/get?secid=0.${ticker}&fields=f58`
    const res = await fetch(url, { signal: AbortSignal.timeout(5000) })
    const data = await res.json() as any

    if (data.data?.f58) {
      return {
        valid: true,
        normalized,
        name: data.data.f58,
        error: '',
      }
    }
  } catch {
    // API call failed
  }

  // Eastmoney API may be unstable, try Sina Finance as fallback
  try {
    // Determine Shanghai or Shenzhen
    const prefix = ticker.substring(0, 1)
    const marketCode = ['6'].includes(prefix) ? 'sh' : 'sz'
    const url = `https://hq.sinajs.cn/list=${marketCode}${ticker}`

    const res = await fetch(url, {
      signal: AbortSignal.timeout(5000),
      headers: {
        'Referer': 'https://finance.sina.com.cn',
      },
    })
    const text = await res.text()

    // Parse Sina Finance response format: var hq_str_sh600519="Kweichow Moutai,..."
    const match = text.match(/="([^"]+)"/)
    if (match && match[1]) {
      const parts = match[1].split(',')
      if (parts.length > 0 && parts[0]) {
        return {
          valid: true,
          normalized,
          name: parts[0],
          error: '',
        }
      }
    }
  } catch {
    // Sina API also failed
  }

  // If format is correct but name unavailable, still consider valid
  return {
    valid: true,
    normalized,
    name: '',
    error: '',
  }
}

/**
 * Validate US/HK stock code and get name
 * Uses Yahoo Finance API
 */
async function validateUSHK(ticker: string, market: Market): Promise<ValidationResult> {
  const normalized = normalizeTicker(ticker, market)

  try {
    const url = `https://query1.finance.yahoo.com/v1/finance/search?q=${encodeURIComponent(normalized)}&quotesCount=5&newsCount=0`
    const res = await fetch(url, { signal: AbortSignal.timeout(5000) })
    const data = await res.json() as any

    if (data.quotes && data.quotes.length > 0) {
      // Find exact match
      const exactMatch = data.quotes.find((q: any) =>
        q.symbol?.toUpperCase() === normalized.toUpperCase()
      )

      if (exactMatch) {
        return {
          valid: true,
          normalized,
          name: exactMatch.shortname || exactMatch.longname || '',
          error: '',
        }
      }

      // If no exact match, use first result
      const first = data.quotes[0]
      return {
        valid: true,
        normalized,
        name: first.shortname || first.longname || '',
        error: '',
      }
    }
  } catch (e) {
    // API call failed
    console.error('Yahoo Finance API error:', e)
  }

  // If format is correct but name unavailable, still consider valid
  return {
    valid: true,
    normalized,
    name: '',
    error: '',
  }
}

/**
 * Validate ticker format only (no name lookup, fast)
 */
export function validateTickerFormat(ticker: string, market: Market): ValidationResult {
  ticker = ticker.trim()

  if (!ticker) {
    return { valid: false, normalized: '', name: '', error: 'Please enter a stock code' }
  }

  if (market === 'A-share') {
    if (!validateAShareFormat(ticker)) {
      return {
        valid: false,
        normalized: '',
        name: '',
        error: 'A-Share code should be 6 digits (e.g. 600519, 000001)',
      }
    }
    return {
      valid: true,
      normalized: normalizeTicker(ticker, market),
      name: '',
      error: '',
    }
  }

  if (market === 'US') {
    if (!validateUSFormat(ticker)) {
      return {
        valid: false,
        normalized: '',
        name: '',
        error: 'US stock code should be 1-5 letters (e.g. AAPL, TSLA)',
      }
    }
    return {
      valid: true,
      normalized: normalizeTicker(ticker, market),
      name: '',
      error: '',
    }
  }

  if (market === 'HK') {
    if (!validateHKFormat(ticker)) {
      return {
        valid: false,
        normalized: '',
        name: '',
        error: 'HK stock code should be 4-5 digits (e.g. 0700, 9988)',
      }
    }
    return {
      valid: true,
      normalized: normalizeTicker(ticker, market),
      name: '',
      error: '',
    }
  }

  return {
    valid: false,
    normalized: '',
    name: '',
    error: `Unsupported market type: ${market}`,
  }
}

/**
 * Validate ticker and get stock name
 */
export async function validateAndGetName(ticker: string, market: Market): Promise<ValidationResult> {
  // Format validation first
  const formatResult = validateTickerFormat(ticker, market)
  if (!formatResult.valid) {
    return formatResult
  }

  const normalized = formatResult.normalized

  // Call different APIs based on market
  if (market === 'A-share') {
    return validateAShare(normalized)
  } else if (market === 'US' || market === 'HK') {
    return validateUSHK(normalized, market)
  }

  return formatResult
}
